### PR TITLE
Version 2.4.0

### PR DESCRIPTION
**Pull request by mpeyfuss**
Update to stacks to include free mint splitting and referral rewards

### CODE REVIEW

File: src/TLStacks1155.sol [hunk 1]
```diff
@@ -577,26 +610,38 @@ contract TLStacks1155 is
...
-            refundAmount = msg.value - totalProtocolFee;
+            if (msg.value != totalProtocolFee) revert IncorrectFunds();
+            if (totalSale > 0) {
+                _safeTransferFromERC20(msg.sender, drop.payoutReceiver, drop.currencyAddress, totalSale);
+            }
+        }
+        uint256 protocolFeeRemaining = totalProtocolFee;
+        // split protocol fee if free mint
+        if (cost == 0) {
+            uint256 split = totalProtocolFee * freeMintFeeSplit / BASIS;
```
Comment (koloz193) [inline src/TLStacks1155.sol:633]: nit: would use a better name

File: src/TLStacks1155.sol [hunk 2]
```diff
@@ -477,6 +500,16 @@ contract TLStacks1155 is
         emit ProtocolFeeUpdated(newProtocolFeeReceiver, newProtocolFee);
     }
 
+    /// @notice Internal function to update the protocol fee split settings
+    /// @param newFreeMintFeeSplit The new fee split for free mints
+    /// @param newReferralFeeSplit The new fee split for referrals
+    function _setProtocolFeeSplits(uint256 newFreeMintFeeSplit, uint256 newReferralFeeSplit) internal {
+        freeMintFeeSplit = newFreeMintFeeSplit;
+        referralFeeSplit = newReferralFeeSplit;
```
Comment (koloz193) [inline src/TLStacks1155.sol:508]: might want to limit the amount being set

File: src/TLStacks721.sol [hunk 3]
```diff
@@ -434,6 +457,16 @@ contract TLStacks721 is
         emit ProtocolFeeUpdated(newProtocolFeeReceiver, newProtocolFee);
     }
 
+    /// @notice Internal function to update the protocol fee split settings
+    /// @param newFreeMintFeeSplit The new fee split for free mints
+    /// @param newReferralFeeSplit The new fee split for referrals
+    function _setProtocolFeeSplits(uint256 newFreeMintFeeSplit, uint256 newReferralFeeSplit) internal {
+        freeMintFeeSplit = newFreeMintFeeSplit;
+        referralFeeSplit = newReferralFeeSplit;
```
Comment (koloz193) [inline src/TLStacks721.sol:465]: same comment as erc1155

File: src/TLStacks1155.sol [hunk 4]
```diff
@@ -577,26 +610,38 @@ contract TLStacks1155 is
...
+            uint256 split = totalProtocolFee * freeMintFeeSplit / BASIS;
+            _safeTransferETH(drop.payoutReceiver, split, weth);
+            protocolFeeRemaining -= split;
         }
-        _safeTransferETH(protocolFeeReceiver, totalProtocolFee, weth);
-        if (refundAmount > 0) {
-            _safeTransferETH(msg.sender, refundAmount, weth);
+        // split protoocl fee if referred
+        if (referralAddress != address(0)) {
+            uint256 split = totalProtocolFee * referralFeeSplit / BASIS;
```
Comment (koloz193) [inline src/TLStacks1155.sol:639]: nit: better name

File: src/TLStacks721.sol [hunk 5]
```diff
@@ -532,26 +565,38 @@ contract TLStacks721 is
...
+            uint256 split = totalProtocolFee * freeMintFeeSplit / BASIS;
+            _safeTransferETH(drop.payoutReceiver, split, weth);
+            protocolFeeRemaining -= split;
         }
-        _safeTransferETH(protocolFeeReceiver, totalProtocolFee, weth);
-        if (refundAmount > 0) {
-            _safeTransferETH(msg.sender, refundAmount, weth);
+        // split protoocl fee if referred
+        if (referralAddress != address(0)) {
+            uint256 split = totalProtocolFee * referralFeeSplit / BASIS;
```
Comment (koloz193) [inline src/TLStacks721.sol:594]: nit: better name

File: src/TLStacks721.sol [hunk 6]
```diff
@@ -532,26 +565,38 @@ contract TLStacks721 is
...
-            refundAmount = msg.value - totalProtocolFee;
+            if (msg.value != totalProtocolFee) revert IncorrectFunds();
+            if (totalSale > 0) {
+                _safeTransferFromERC20(msg.sender, drop.payoutReceiver, drop.currencyAddress, totalSale);
+            }
+        }
+        uint256 protocolFeeRemaining = totalProtocolFee;
+        // split protocol fee if free mint
+        if (cost == 0) {
+            uint256 split = totalProtocolFee * freeMintFeeSplit / BASIS;
```
Comment (koloz193) [inline src/TLStacks721.sol:588]: nit: better name

File: src/TLStacks1155.sol (hunk 2 above)
Comment (mpeyfuss) [reply src/TLStacks1155.sol:508]: good catch

**Review by koloz193 — APPROVED** [review]

**Pull request merged**